### PR TITLE
GRF-541 : use category template and attribute  database information 

### PR DESCRIPTION
**Pull request by antoinetrouve**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**
[screen-capture.webm](https://user-images.githubusercontent.com/6566364/200298169-569a715c-abd8-44ce-85f9-5ab90466b9a8.webm)

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Category/back/Domain/Model/Enrichment/Category.php [hunk 1]
```diff
@@ -35,25 +37,27 @@ public function __construct(
...
      *     root_id: int|null,
      *     value_collection: string|null,
-     *     permissions: string|null
-     * } $result
+     *     permissions: string|null,
+     *     template_uuid: string|null
+     * } $category
      */
-    public static function fromDatabase(array $result): self
+    public static function fromDatabase(array $category): self
```
Comment (jeremyBeaucousinAkeneo) [inline src/Akeneo/Category/back/Domain/Model/Enrichment/Category.php:44]: I juste wandered why rename it. It seems raw data in argument. Juste asking

File: src/Akeneo/Category/back/Domain/ValueObject/Attribute/AttributeCollection.php [hunk 2]
```diff
@@ -92,4 +92,9 @@ public function normalize(): array
             $this->attributes,
         );
     }
+
+    public function count(): int
+    {
+        return count($this->attributes);
+    }
```
Comment (jeremyBeaucousinAkeneo) [inline src/Akeneo/Category/back/Domain/ValueObject/Attribute/AttributeCollection.php:99]: I cannot see the advantage of wrapping a nativ function into a method

File: src/Akeneo/Category/back/Infrastructure/Storage/InMemory/GetAttributeInMemory.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+use Akeneo\Category\Domain\ValueObject\Attribute\AttributeOrder;
+use Akeneo\Category\Domain\ValueObject\Attribute\AttributeUuid;
+use Akeneo\Category\Domain\ValueObject\LabelCollection;
+use Akeneo\Category\Domain\ValueObject\Template\TemplateUuid;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class GetAttributeInMemory implements GetAttribute
```
Comment (jeremyBeaucousinAkeneo) [inline src/Akeneo/Category/back/Infrastructure/Storage/InMemory/GetAttributeInMemory.php:28]: It seems it's only for test (if I'm not mistaking). In that case should we move it ?

**Review by jeremyBeaucousinAkeneo — APPROVED** [review]

File: src/Akeneo/Category/back/Domain/Model/Enrichment/Category.php (hunk 1 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Domain/Model/Enrichment/Category.php:44]: For me, I think it's better to understand and more functional speaking.

File: src/Akeneo/Category/back/Domain/ValueObject/Attribute/AttributeCollection.php (hunk 2 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Domain/ValueObject/Attribute/AttributeCollection.php:99]: `count() `can be used both on arrays and Countable objects. If you are using $object->count() you are restricting your function to just objects. Using count() you allow arrays too.

See here :  https://github.com/php/php-src/blob/master/ext/standard/array.c#L676

File: src/Akeneo/Category/back/Infrastructure/Storage/InMemory/GetAttributeInMemory.php (hunk 3 above)
Comment (antoinetrouve) [reply src/Akeneo/Category/back/Infrastructure/Storage/InMemory/GetAttributeInMemory.php:28]: Yes for now, but we decided to keep this class if needed and simulate retrieving data from database. This class must have the same behavior than the GetAttributeSql to simulate its use.

**Pull request merged**